File: backend/app/schemas/machinery.py
"""Schemas Pydantic para maquinaria, mantenimiento y uso."""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ── Machine ───────────────────────────────────────────

class MachineCreate(BaseModel):
    code: str = Field(min_length=1, max_length=50)
    name: str = Field(min_length=2, max_length=255)
    type: str
    brand: str = Field(max_length=100)
    model: str = Field(max_length=100)
    year: int = Field(ge=1970, le=2100)
    serial_number: str = Field(max_length=100)
    capacity: Optional[str] = None
    current_location: Optional[str] = None
    status: str = "disponible"
    hourometer_current: Decimal = Decimal("0")
    hourometer_next_service: Decimal = Decimal("500")
    daily_cost: Decimal = Decimal("0")
    acquisition_cost: Decimal = Decimal("0")
    image_url: Optional[str] = None
    notes: Optional[str] = None


class MachineUpdate(BaseModel):
    name: Optional[str] = None
    type: Optional[str] = None
    brand: Optional[str] = None
    model: Optional[str] = None
    year: Optional[int] = None
    serial_number: Optional[str] = None
    capacity: Optional[str] = None
    current_location: Optional[str] = None
    status: Optional[str] = None
    hourometer_current: Optional[Decimal] = None
    hourometer_next_service: Optional[Decimal] = None
    daily_cost: Optional[Decimal] = None
    acquisition_cost: Optional[Decimal] = None
    image_url: Optional[str] = None
    notes: Optional[str] = None


class MachineResponse(BaseModel):
    id: UUID
    code: str
    name: str
    type: str
    brand: str
    model: str
    year: int
    serial_number: str
    capacity: Optional[str] = None
    current_location: Optional[str] = None
    status: str
    hourometer_current: Decimal
    hourometer_next_service: Decimal
    hours_until_service: Decimal
    daily_cost: Decimal
    acquisition_cost: Decimal
    image_url: Optional[str] = None
    notes: Optional[str] = None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class MachineListResponse(BaseModel):
    id: UUID
    code: str
    name: str
    type: str
    brand: str
    status: str
    hourometer_current: Decimal
    hourometer_next_service: Decimal
    hours_until_service: Decimal
    current_location: Optional[str] = None
    daily_cost: Decimal

    model_config = {"from_attributes": True}


# ── MaintenanceLog ────────────────────────────────────

class MaintenanceLogCreate(BaseModel):
    type: str  # preventivo, correctivo, emergencia
    description: str = Field(min_length=5)
    cost: Decimal = Decimal("0")
    parts_used: Optional[list[dict]] = None  # [{"part": "Filtro", "qty": 2, "cost": 350}]
    performed_by: str = Field(min_length=2)
    performed_at: datetime
    hourometer_at_service: Decimal
    downtime_hours: Decimal = Decimal("0")
    next_maintenance_hours: Optional[Decimal] = None


class MaintenanceLogResponse(BaseModel):
    id: UUID
    machine_id: UUID
    type: str
    description: str
    cost: Decimal
    parts_used: Optional[list[dict]] = None
    performed_by: str
    performed_at: datetime
    hourometer_at_service: Decimal
    downtime_hours: Decimal
    next_maintenance_hours: Optional[Decimal] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── UsageLog ──────────────────────────────────────────

class UsageLogCreate(BaseModel):
    project_id: Optional[UUID] = None
    date: date
    hours_worked: Decimal = Field(gt=0)
    hourometer_start: Decimal
    hourometer_end: Decimal
    operator: str = Field(min_length=2)
    fuel_liters: Optional[Decimal] = None
    notes: Optional[str] = None


class UsageLogResponse(BaseModel):
    id: UUID
    machine_id: UUID
    project_id: Optional[UUID] = None
    date: date
    hours_worked: Decimal
    hourometer_start: Decimal
    hourometer_end: Decimal
    operator: str
    fuel_liters: Optional[Decimal] = None
    notes: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/machinery.ts
import client from "./client";

export interface Machine {
  id: string;
  code: string;
  name: string;
  type: string;
  brand: string;
  model?: string;
  year?: number;
  serial_number?: string;
  capacity?: string;
  current_location?: string;
  status: string;
  hourometer_current: number;
  hourometer_next_service: number;
  hours_until_service: number;
  daily_cost: number;
  acquisition_cost?: number;
  image_url?: string;
  notes?: string;
  is_active: boolean;
  created_at: string;
}

export interface MachineCreate {
  code: string;
  name: string;
  type: string;
  brand: string;
  model: string;
  year: number;
  serial_number: string;
  capacity?: string;
  current_location?: string;
  status?: string;
  hourometer_current?: number;
  hourometer_next_service?: number;
  daily_cost?: number;
  acquisition_cost?: number;
  notes?: string;
}

export interface UsageLog {
  id: string;
  machine_id: string;
  project_id?: string;
  date: string;
  hours_worked: number;
  hourometer_start: number;
  hourometer_end: number;
  operator: string;
  fuel_liters?: number;
  notes?: string;
  created_at: string;
}

export interface UsageLogCreate {
  project_id?: string;
  date: string;
  hours_worked: number;
  hourometer_start: number;
  hourometer_end: number;
  operator: string;
  fuel_liters?: number;
  notes?: string;
}

export interface MaintenanceLog {
  id: string;
  machine_id: string;
  type: string;
  description: string;
  cost: number;
  parts_used?: { part: string; qty: number; cost: number }[];
  performed_by: string;
  performed_at: string;
  hourometer_at_service: number;
  downtime_hours: number;
  next_maintenance_hours?: number;
  created_at: string;
}

export interface MaintenanceLogCreate {
  type: string;
  description: string;
  cost?: number;
  parts_used?: { part: string; qty: number; cost: number }[];
  performed_by: string;
  performed_at: string;
  hourometer_at_service: number;
  downtime_hours?: number;
  next_maintenance_hours?: number;
}

export interface MachineStats {
  usage: {
    period_days: number;
    total_entries: number;
    total_hours: number;
    total_fuel: number;
    avg_daily_hours: number;
  };
  daily_hours: { date: string; hours: number }[];
  maintenance: {
    by_type: Record<string, { count: number; total_cost: number }>;
    total_cost: number;
  };
}

export const machineryApi = {
  list: (params?: { status?: string; type?: string; search?: string }) =>
    client.get<Machine[]>("/machinery", { params }),

  get: (id: string) =>
    client.get<Machine>(`/machinery/${id}`),

  create: (data: MachineCreate) =>
    client.post<Machine>("/machinery", data),

  update: (id: string, data: Partial<MachineCreate>) =>
    client.put<Machine>(`/machinery/${id}`, data),

  delete: (id: string) =>
    client.delete(`/machinery/${id}`),

  alerts: () =>
    client.get<Machine[]>("/machinery/alerts"),

  // Uso
  registerUsage: (machineId: string, data: UsageLogCreate) =>
    client.post<UsageLog>(`/machinery/${machineId}/usage`, data),

  getUsageLogs: (machineId: string, params?: { skip?: number; limit?: number }) =>
    client.get<UsageLog[]>(`/machinery/${machineId}/logs`, { params }),

  getStats: (machineId: string, days = 30) =>
    client.get<MachineStats>(`/machinery/${machineId}/stats`, { params: { days } }),

  // Mantenimiento
  registerMaintenance: (machineId: string, data: MaintenanceLogCreate) =>
    client.post<MaintenanceLog>(`/machinery/${machineId}/maintenance`, data),

  getMaintenanceLogs: (machineId: string, params?: { type?: string }) =>
    client.get<MaintenanceLog[]>(`/machinery/${machineId}/maintenance`, { params }),
};
